File: app/src/main/java/com/example/ta_avance/actividades/ReporteActivity.java
package com.example.ta_avance.actividades;

import android.app.DatePickerDialog;
import android.os.Bundle;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;
import androidx.cardview.widget.CardView;
import androidx.lifecycle.ViewModelProvider;

import com.example.ta_avance.R;
import com.example.ta_avance.dto.servicio.ServicioDto;
import com.example.ta_avance.viewmodel.GestionarServicioViewModel;
import com.example.ta_avance.viewmodel.ReporteViewModel;
import com.google.android.material.textfield.TextInputEditText;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class ReporteActivity extends AppCompatActivity {

    private TextInputEditText etFechaInicio, etFechaFin;
    private Spinner spinnerServicio;
    private Button btnFiltrarReporte;
    private TextView tvServicioNombre, tvMontoTotal, tvCantidadReservas;
    private CardView cardResultado;

    private GestionarServicioViewModel servicioViewModel;
    private ReporteViewModel reporteViewModel;

    private final List<ServicioDto> listaServicios = new ArrayList<>();
    private LocalDate fechaInicioSeleccionada, fechaFinSeleccionada;
    private final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        setContentView(R.layout.activity_reporte);

        etFechaInicio = findViewById(R.id.etFechaInicio);
        etFechaFin = findViewById(R.id.etFechaFin);
        spinnerServicio = findViewById(R.id.spinnerServicio);
        btnFiltrarReporte = findViewById(R.id.btnFiltrarReporte);
        tvServicioNombre = findViewById(R.id.tvServicioNombre);
        tvMontoTotal = findViewById(R.id.tvMontoTotal);
        tvCantidadReservas = findViewById(R.id.tvCantidadReservas);
        cardResultado = findViewById(R.id.cardResultado);

        servicioViewModel = new ViewModelProvider(this).get(GestionarServicioViewModel.class);
        reporteViewModel = new ViewModelProvider(this).get(ReporteViewModel.class);

        servicioViewModel.getServicios().observe(this, servicios -> {
            listaServicios.clear();
            listaServicios.addAll(servicios);

            List<String> nombres = new ArrayList<>();
            nombres.add("Todos");
            for (ServicioDto s : servicios) nombres.add(s.getNombre());

            ArrayAdapter<String> adapter = new ArrayAdapter<>(this,
                    android.R.layout.simple_spinner_dropdown_item, nombres);
            spinnerServicio.setAdapter(adapter);
        });

        reporteViewModel.getReporte().observe(this, reporte -> {
            tvServicioNombre.setText(reporte.getServicioNombre() != null ? reporte.getServicioNombre() : "Todos");
            tvMontoTotal.setText("S/ " + reporte.getMontoTotal());
            tvCantidadReservas.setText("" + reporte.getCantidadReservas());
            cardResultado.setVisibility(View.VISIBLE);
        });

        reporteViewModel.getError().observe(this, msg ->
                Toast.makeText(this, msg, Toast.LENGTH_SHORT).show()
        );

        servicioViewModel.getError().observe(this, msg ->
                Toast.makeText(this, msg, Toast.LENGTH_SHORT).show()
        );

        etFechaInicio.setOnClickListener(v -> mostrarDatePicker(true));
        etFechaFin.setOnClickListener(v -> mostrarDatePicker(false));

        cargarServicios();
        btnFiltrarReporte.setOnClickListener(v -> filtrarReporte());
    }

    private void mostrarDatePicker(boolean esInicio) {
        final Calendar cal = Calendar.getInstance();
        DatePickerDialog dp = new DatePickerDialog(this, (view, y, m, d) -> {
            LocalDate fecha = LocalDate.of(y, m + 1, d);
            if (esInicio) {
                fechaInicioSeleccionada = fecha;
                etFechaInicio.setText(formatter.format(fecha));
            } else {
                fechaFinSeleccionada = fecha;
                etFechaFin.setText(formatter.format(fecha));
            }
        }, cal.get(Calendar.YEAR), cal.get(Calendar.MONTH), cal.get(Calendar.DAY_OF_MONTH));
        dp.show();
    }

    private void cargarServicios() {
        servicioViewModel.obtenerServicios();
    }

    private void filtrarReporte() {
        if (fechaInicioSeleccionada == null || fechaFinSeleccionada == null) {
            Toast.makeText(this, "Selecciona ambas fechas", Toast.LENGTH_SHORT).show();
            return;
        }

        String servicioSeleccionado;
        int pos = spinnerServicio.getSelectedItemPosition();
        if (pos == 0) {
            servicioSeleccionado = null;
        } else {
            servicioSeleccionado = listaServicios.get(pos - 1).getNombre();
        }

        cardResultado.setVisibility(View.GONE);
        reporteViewModel.obtenerReporte(fechaInicioSeleccionada, fechaFinSeleccionada, servicioSeleccionado);
    }
}

File: app/src/main/java/com/example/ta_avance/actividades/ReporteActivity.kt
package com.example.ta_avance.actividades

import android.app.DatePickerDialog
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.ArrayAdapter
import android.widget.Button
import android.widget.Spinner
import android.widget.TextView
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.cardview.widget.CardView
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.example.ta_avance.R
import com.shared.models.dto.servicio.ServicioDto
import com.shared.models.ui.state.UiState
import com.example.ta_avance.viewmodel.GestionarServicioViewModel
import com.example.ta_avance.viewmodel.ReporteViewModel
import com.google.android.material.button.MaterialButton
import com.google.android.material.textfield.TextInputEditText
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import java.io.File
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.util.Calendar

@AndroidEntryPoint
class ReporteActivity : AppCompatActivity() {

    private lateinit var etFechaInicio: TextInputEditText
    private lateinit var etFechaFin: TextInputEditText
    private lateinit var spinnerServicio: Spinner
    private lateinit var tvServicioNombre: TextView
    private lateinit var tvMontoTotal: TextView
    private lateinit var tvCantidadReservas: TextView
    private lateinit var cardResultado: CardView
    private lateinit var btnDescargarPdf: MaterialButton
    private lateinit var btnCompartirPdf: MaterialButton
    private lateinit var servicioViewModel: GestionarServicioViewModel
    private lateinit var reporteViewModel: ReporteViewModel

    private val listaServicios = mutableListOf<ServicioDto>()
    private var fechaInicioSeleccionada: LocalDate? = null
    private var fechaFinSeleccionada: LocalDate? = null
    private val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd")
    private var archivoPdf: File? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        setContentView(R.layout.activity_reporte)

        etFechaInicio = findViewById(R.id.etFechaInicio)
        etFechaFin = findViewById(R.id.etFechaFin)
        spinnerServicio = findViewById(R.id.spinnerServicio)
        tvServicioNombre = findViewById(R.id.tvServicioNombre)
        tvMontoTotal = findViewById(R.id.tvMontoTotal)
        tvCantidadReservas = findViewById(R.id.tvCantidadReservas)
        cardResultado = findViewById(R.id.cardResultado)
        btnDescargarPdf = findViewById(R.id.btnDescargarPdf)
        btnCompartirPdf = findViewById(R.id.btnCompartirPdf)

        servicioViewModel = ViewModelProvider(this).get(GestionarServicioViewModel::class.java)
        reporteViewModel = ViewModelProvider(this).get(ReporteViewModel::class.java)

        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                servicioViewModel.listState.collect { state ->
                    when (state) {
                        is UiState.Success -> {
                            listaServicios.clear()
                            listaServicios.addAll(state.data)
                            val nombres = reporteViewModel.generarNombresServicios(state.data)
                            spinnerServicio.adapter = ArrayAdapter(this@ReporteActivity, android.R.layout.simple_spinner_dropdown_item, nombres)
                        }
                        is UiState.Error -> Toast.makeText(this@ReporteActivity, state.message, Toast.LENGTH_SHORT).show()
                        else -> {}
                    }
                }
            }
        }

        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                reporteViewModel.reporteState.collect { state ->
                    when (state) {
                        is UiState.Success -> {
                            val reporte = state.data
                            tvServicioNombre.text = reporte.servicioNombre ?: "Todos"
                            tvMontoTotal.text = "S/ ${reporte.montoTotal}"
                            tvCantidadReservas.text = "${reporte.cantidadReservas}"
                            cardResultado.visibility = View.VISIBLE
                            btnDescargarPdf.visibility = View.VISIBLE
                        }
                        is UiState.Error -> Toast.makeText(this@ReporteActivity, state.message, Toast.LENGTH_SHORT).show()
                        else -> {}
                    }
                }
            }
        }

        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                reporteViewModel.descargaState.collect { state ->
                    when (state) {
                        is UiState.Loading -> {
                            btnDescargarPdf.isEnabled = false
                            btnDescargarPdf.text = "Descargando..."
                        }
                        is UiState.Success -> {
                            archivoPdf = state.data
                            btnDescargarPdf.isEnabled = true
                            btnDescargarPdf.text = "Descargar PDF"
                            btnCompartirPdf.visibility = View.VISIBLE
                            Toast.makeText(this@ReporteActivity, "PDF guardado en Descargas", Toast.LENGTH_SHORT).show()
                        }
                        is UiState.Error -> {
                            btnDescargarPdf.isEnabled = true
                            btnDescargarPdf.text = "Descargar PDF"
                            Toast.makeText(this@ReporteActivity, state.message, Toast.LENGTH_SHORT).show()
                        }
                        else -> {}
                    }
                }
            }
        }

        etFechaInicio.setOnClickListener { mostrarDatePicker(true) }
        etFechaFin.setOnClickListener { mostrarDatePicker(false) }
        findViewById<Button>(R.id.btnFiltrarReporte).setOnClickListener { filtrarReporte() }

        btnDescargarPdf.setOnClickListener {
            reporteViewModel.descargarReporte(this, fechaInicioSeleccionada, fechaFinSeleccionada)
        }

        btnCompartirPdf.setOnClickListener {
            val archivo = archivoPdf ?: return@setOnClickListener
            val intent = reporteViewModel.construirIntentCompartir(this, archivo)
            startActivity(Intent.createChooser(intent, "Compartir reporte"))
        }

        servicioViewModel.obtenerServicios()
    }

    private fun mostrarDatePicker(esInicio: Boolean) {
        val cal = Calendar.getInstance()
        DatePickerDialog(this, { _, y, m, d ->
            val fecha = LocalDate.of(y, m + 1, d)
            if (esInicio) {
                fechaInicioSeleccionada = fecha
                etFechaInicio.setText(formatter.format(fecha))
            } else {
                fechaFinSeleccionada = fecha
                etFechaFin.setText(formatter.format(fecha))
            }
        }, cal[Calendar.YEAR], cal[Calendar.MONTH], cal[Calendar.DAY_OF_MONTH]).show()
    }

    private fun filtrarReporte() {
        cardResultado.visibility = View.GONE
        btnDescargarPdf.visibility = View.GONE
        btnCompartirPdf.visibility = View.GONE
        archivoPdf = null
        val servicioSeleccionado = reporteViewModel.resolverServicioSeleccionado(
            listaServicios,
            spinnerServicio.selectedItemPosition
        )
        reporteViewModel.obtenerReporte(fechaInicioSeleccionada, fechaFinSeleccionada, servicioSeleccionado)
    }
}
